sno/checkout.py

```diff
@@ -307,6 +307,41 @@ def restore(ctx, source, pathspec):
     )
 
 
+@click.command()
+@click.pass_context
+@click.option(
+    "--discard-changes",
+    "--force",
+    "-f",
+    is_flag=True,
+    help="Discard local changes in working copy if necessary",
+)
+@click.argument("refish")
+def reset(ctx, discard_changes, refish):
```

Comment:
So this command is similar to `git reset` but has some differences:

* `sno reset <commit>` - is like `git reset --keep <commit>` - it resets the working copy as well as the branch tip, but refuses to discard local changes 👍 

* `sno reset --discard-changes <commit>` is like `git reset --hard <commit>` - it resets the working copy as well as the branch tip, _and_ discards local changes. This option might not actually be necessary since you can always `sno restore` to discard the local changes before running `sno reset`. I'd tend toward leaving this option out.

I feel like sooner or later we're going to want a soft-reset option to allow the user to leave the working copy alone `--mixed` is the default in `git reset` and comes in handy for some workflows. Probably a more descriptive name wouldn't hurt, e.g. `--ignore-working-copy`. We don't need to distinguish between 'soft' and 'mixed' since we don't have an index area.

@UberMouse mentioned he would find a soft-reset option useful ( https://github.com/koordinates/sno/issues/60#issuecomment-618246032 )

Reply:
I don't necessarily think of it as a shortcut, but I think `--discard-changes` should stay because
a) it is consistent with existing sno commands, lots of which won't run if there's changes unless you --discard-changes.
b) It's useful to put in error messages. You just told sno to do something but it didn't due to uncommitted changes. But, it suggests that if you run the command again with --discard-changes, then it will do it. Anything more than that - even if it's just "run sno restore first" - seems to me like we are sending the user on a slight detour (even though the total characters typed is no worse) - Sno won't run THIS command unless you first run THAT command, and it's less clear from the name what THAT command does.

I talked to @UberMouse today and he was pretty vague about what he once needed this for, which is unsurprising for a FR that is unimplemented for 6 months - he must have found a workaround. He told me that he uses these things only for dev - his requiring them or not isn't a good indicator if they are generally useful - and right now he was pretty ¯\_(ツ)_/¯ about if a --hard version would be okay for him instead of --soft.

There's a bunch more context at the bottom of #60 - 
- Rob has already okayed this and the removal of the current `sno reset`
- I explain how we don't support mixed in general - that is, any command that updates your HEAD currently doesn't let you do so while keeping uncommitted working copy changes. This could be fixed but it's a more in depth change, and a bit different for us than it is in git (git can't get schema violations by doing a mixed reset, but we can)

Reply:
okay. I expect someone'll request a soft-reset eventually, but until then we can ignore it